docs/src/developer/future.md

```diff
@@ -0,0 +1,126 @@
+```@meta
+CurrentModule = Nemo
+```
+
+# Future plans
+
+## Ring and CommRing
+
+Currently all commutative ring types belong to `Ring` and their elements to
+`RingElem` (and `RingElement`) and we have separate types for noncommutative
+rings and elements thereof, i.e. `NCRing` and `NCRingElem` etc.
+
+However, it would be more logical to use `Ring` for not necessarily
+commutative rings and `CommRing`, `CRing` or `CommutativeRing` (the name has
+not been decided on yet) for commutative rings.
+
+This is a big change and should happen with plenty of warning for the
+community. It would be convenient if a script could be made available to
+automate this.
+
+## Matrices and polynomials over number fields
+
+The Antic C library is used for number field elements, but matrices and
+polynomials over these are handled by Julia. There is nothing explicitly
+wrong with this, except that the jit compilation is costly for such basic
+operations and the code is not as easy for others to use in other projects
+if it is not in C.
+
+We plan to write some routines in C in the Antic library to handle at least
+univariate polynomials over number fields and matrices as well. These should
+use highly optimised routines based on the algorithms with best complexity
+for these operations.
+
+## Mono repository
+
+There is currently a proposal to place all Oscar related repositories, or some
+subset of them in a single repository called OscarMono.jl. The details are not
+finalised and it is not known what impact this will have on Nemo. However,
+Nemo developers should be aware that this may happen at some point in the
+fairly near future.
+
+Users of Nemo should be unaffected, as Nemo will continue to exist as a
+separate package in the OscarMono.jl repository, even if it does become part of
+this repository. Julia apparently supports multiple packages in the same
+repository nowadays.
+
+The possibility will always exist to separate the repositories again if the
+experiment is unsuccessful or serves its purpose and is no longer needed.
+
+## Splitting abstract and generic functionality
+
+Currently we have many generic types in Nemo, e.g. `Poly{T}` and `Mat{T}`. We
+also have a lot of generic algorithms implemented. The latter are implemented
+for abstract types.
+
+At the present moment, both the specialised code for our generic types and the
+generic algorithms for abstract types are all implemented inside the
+`src/generic` directory of the AbstractAlgebra package.
+
+We eventually plan to separate the two kinds of implementations to make it
+clearer to developers which code is for generic types only and which are the
+more general implementations for all types belonging to the AbstractAlgebra
+abstract types.
+
+It's not clear at this moment what the new directories will be called when the
+separation happens.
```

Comment:
Side note: some development started to happen in `src/algorithms.jl` for this.

Reply:
Right, but all of that code is not going to fit in a single file, so I'll leave out mention of this for now. We can add it later when the decision becomes canonical.

Reply:
Oh right, and I meant `src/algorithm/` which is actually a directory. But totally fair point about canonicalising the decision first, I didn't mean to ask for a change here.